File: TTT_VS_AI.py
from termcolor import colored, cprint
import random

def main():
    global a

    per = 0
    m = "start"
    massage = "oopss"
    print(colored("Выберете сложность","magenta",attrs=['bold']))
    print((colored("Для того чтобы выбрать 'Лёгко' введите [0]","magenta",attrs=['bold'])))
    print((colored("Для того чтобы выбрать 'Сложно' введите [1]","magenta",attrs=['bold'])))
    level = int(input(""))
    if level == 0: massage = "Лёгкий режим"
    if level == 1: massage = "Сложный режим"
    print((colored("Вы выбрали","green",attrs=['bold'])),colored(massage,"magenta",attrs=['bold']),"\n")
    print(colored("первыми ходят x","magenta",attrs=['bold']))
    print((colored("Для того чтобы выбрать x введите [1]","magenta",attrs=['bold'])))
    print((colored("Для того чтобы выбрать o введите [2]","magenta",attrs=['bold'])))
    n = int(input(""))
    if n == 1: 
        m = "x"
        per += 1
    if n == 2: m = "o"
    print((colored("Вы выбрали","green",attrs=['bold'])),colored(m,"magenta",attrs=['bold']),"\n")
    a = [["_"]*3 for i in range(3)]
    for i in a:
        print(*i)
    
    flag = True
    br = True
    while flag and br:
        br = False
        sub = True
        if per%2 == 0:
            cprint('----Ход ИИ----','blue') 
            AI_move(per,m,level)
        if per%2 == 1:
            while sub:
                print("Куда хотите поставить",m,':')
                i,j = map(int, input().split())
                i -= 1
                j -= 1
                if a[i][j] == '_':
                    a[i][j] = m
                    sub = False
                else:
                    print(colored('Данное поле уже занято значком:','red'),a[i][j])
        for i in a:
            print(*i)
        state = scan()
        flag = state[0]
        br = state[1]
        win = state[2]
        per += 1
    if br == False and flag != False:
        cprint("Победителя нет",'yellow')
    if flag == False:
        print(colored("Победил",'green'),win)

def scan():
    """
    -Сканирует поле на предмет переполненности или выигрыша
    -Принимает поле(двоичный массив 3x3)
    -Возвращает флаг и победителя, если нашёлся
    """
    win = 0
    br = False
    flag = True
    for i in range(3):
        if scan_line([i,0],[i,1],[i,2]):
            win = a[i][0]
            flag = False
        if scan_line([0,i],[1,i],[2,i]):
            win = a[0][i]
            flag = False
    if scan_line([0,0],[1,1],[2,2]):
        win = a[0][0]
        flag = False
    if scan_line([0,2],[1,1],[2,0]):
        win = a[2][0]
        flag = False
    for i in a:
        for j in i:
            if j == '_':
                br = True
    y = [flag,br,win]
    return y
def scan_line(e1,e2,e3):
    result = False
    if a[e1[0]][e1[1]] == a[e2[0]][e2[1]] and a[e2[0]][e2[1]] == a[e3[0]][e3[1]] and a[e3[0]][e3[1]] != '_':
        result = True
    return result

def AI_move(per:int,m:str,level:int):
    """
    -Обработка поля программой 
    -Принимает шаг
    -Возвращает изменённое поле
    param per: move count
    param m: противник
    param level: уровень игры level 1 - сложный; 0 - лёгкий
    """
    value = "start" #значение ИИ в зависимости от параметра m
    #Первые 2 хода обрабатываем вручную, они ключевые
    if m == "o":
        value = "x"
        if level == 1:
            if per == 0:
                a[0][0] = "x"
                return
            elif per == 2:
                if a[2][0] == '_':
                    a[2][0] = 'x'
                elif a[2][2] == '_':
                    a[2][2] = 'x'
                return
    if m == "x":
        value = "o"
        if level == 1:
            if per == 2:
                if a[1][1] == "_": a[1][1] = "o"
                else: a[0][0] = "o"
                return
    #ходы дальше обрабатываются автоматически
    if (per >= 4 and level == 1) or level == 0:
        #обрабатываем критическую ситуацию(2 value)
        for i in range(3): #проверяем строки
            if scan_to_win([i,0],[i,1],[i,2],value,value):
                return
        for i in range(3): #проверяем столбцы
            if scan_to_win([0,i],[1,i],[2,i],value,value):
                return
        if scan_to_win([0,0],[1,1],[2,2],value,value):#проверяем главную диагональ
            return
        if scan_to_win([0,2],[1,1],[2,0],value,value):#проверяем побочную диагональ
            return

        #обрабатываем критическую ситуацию(2 m)
        for i in range(3): #проверяем строки
            if scan_to_win([i,0],[i,1],[i,2],m,value):
                return
        for i in range(3): #проверяем столбцы
            if scan_to_win([0,i],[1,i],[2,i],m,value):
                return
        if scan_to_win([0,0],[1,1],[2,2],m,value):#проверяем главную диагональ
            return
        if scan_to_win([0,2],[1,1],[2,0],m,value):#проверяем побочную диагональ
            return

        #проверяем случай когда 1 value и остальное пусто
        for i in range(3): #проверяем строки
            if scan_to_move([i,0],[i,1],[i,2],value,value):
                return
        for i in range(3): #проверяем столбцы
            if scan_to_move([0,i],[1,i],[2,i],value,value):
                return
        if scan_to_move([0,0],[1,1],[2,2],value,value):#проверяем главную диагональ
            return
        if scan_to_move([0,2],[1,1],[2,0],value,value):#проверяем побочную диагональ
            return

        #во всех остальных случаях значение ставится рандомно
        while True:
            i = random.randint(0, 2)
            j = random.randint(0, 2)
            if a[i][j] == '_':
                a[i][j] = value
                break


def scan_to_move(e1,e2,e3,s,m):#!!!!!!!!!!!! можно сделать рандом !!!!!!!!!!!!
    """
    -Проверяет ситуацию вида x__
    -Сканирует заданную линию на поле
    -Если ситуация, то делает ход
    -Возвращает флаг 
    -param s: значение которое ищем
    -param m: значение ИИ
    """
    result = False
    if a[e1[0]][e1[1]] == s and a[e2[0]][e2[1]] == '_' and a[e3[0]][e3[1]] == '_':
        a[e3[0]][e3[1]] = m
        result = True
    if a[e1[0]][e1[1]] == '_' and a[e2[0]][e2[1]] == s and a[e3[0]][e3[1]]  == '_':
        a[e3[0]][e3[1]] = m
        result = True
    if a[e1[0]][e1[1]] == '_' and a[e2[0]][e2[1]] == '_' and a[e3[0]][e3[1]] == s:
        a[e1[0]][e1[1]] = m
        result = True
    return result

def scan_to_win(e1,e2,e3,s,m):
    """
    -Проверяет ситуацию вида oo_ или xx_
    -Сканирует заданную линию на поле
    -Если ситуация "критическая", то делает ход
    -Возвращает флаг
    -param s: значение которое ищем
    -param m: значение ИИ 
    """
    result = False
    if a[e1[0]][e1[1]] == s and a[e2[0]][e2[1]] == s and a[e3[0]][e3[1]] == '_':
        a[e3[0]][e3[1]] = m
        result = True
    if a[e1[0]][e1[1]] == s and a[e2[0]][e2[1]] == '_' and a[e3[0]][e3[1]] == s:
        a[e2[0]][e2[1]] = m
        result = True
    if a[e1[0]][e1[1]] == '_' and a[e2[0]][e2[1]] == s and a[e3[0]][e3[1]] == s:
        a[e1[0]][e1[1]] = m
        result = True
    return result

main()

File: TTT_VS_AI.c
#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <stdbool.h>
#include <locale.h>
#include <Windows.h>
#include <time.h>


int scan(bool *,bool *,char *);
bool scan_line(char,char,char);
int AI_move();
int scan_to_move(char *, char *, char *);
int scan_to_win(char *,char *,char *,char);


 
char a[3][3]; /*Поле*/
size_t per=0; /*Ход*/
char m/*соперник*/,value/*ИИ*/;
int level=2;/*уровень игры*/

int main(){
    system("CLS");
    size_t i,j,co[2];
    char win;
    int user_choice = 2;
    bool flag = TRUE,br = TRUE,user_br = TRUE,sub = TRUE;

    if( SetConsoleCP(CP_UTF8) == 0 || SetConsoleOutputCP(CP_UTF8) == 0) printf("Error!\n");

    while(sub){
        printf("Выберете сложность\nДля того чтобы выбрать 'Лёгко' введите [0]\nДля того чтобы выбрать 'Сложно' введите [1]\n");
        scanf("%d",&level);
        fflush(stdin);
        if(level == 1 || level == 0){
            sub = FALSE;
        }
        else{
            printf("Error!--Введено неверное значение.Попробуйте снова.\n\n");
        }
    }
    sub = TRUE;
    if(level == 0) printf("Вы выбрали Лёгкий режим.\n\n");
    if(level == 1) printf("Вы выбрали Сложный режим.\n\n");

    while(sub){
        printf("первыми ходят x\nДля того чтобы выбрать 'x' введите [0]\nДля того чтобы выбрать 'o' введите [1]\n");
        scanf("%d",&user_choice);
        fflush(stdin);
        if(user_choice == 1 || user_choice == 0){
            sub = FALSE;
        }
        else{
            printf("Error!--Введено неверное значение.Попробуйте снова.\n\n");
        } 
    }
    sub = TRUE;
    if(user_choice == 0){
        m = 'x';
        per += 1;
    }
    if(user_choice == 1) m = 'o';
    printf("Вы выбрали '%c'\n\n",m);
    
    printf("Ставьте свой знак по аналогии с примером:\nПоставим 'x' в: 1 1\n");
    for(i=0;i < 3;i++){
        for(j=0;j < 3;j++) a[i][j] = '_';
    }
    a[0][0] = 'x';
    for(i=0;i < 3;i++){
        for(j=0;j < 3;j++) printf(" %c ",a[i][j]);
        printf("\n");
    }
    printf("Поставим 'x' в: 2 3\n");
    a[1][2] = 'x';
    for(i=0;i < 3;i++){
        for(j=0;j < 3;j++) printf(" %c ",a[i][j]);
        printf("\n");
    }
    printf("\n\n");

    for(i=0;i < 3;i++){
        for(j=0;j < 3;j++) a[i][j] = '_';
    }
    for(i=0;i < 3;i++){
        for(j=0;j < 3;j++) printf(" %c ",a[i][j]);
        printf("\n");
    }

    while(flag && br && user_br){
        if(per%2 == 0){
            printf("----Ход ИИ----\n");
            if(AI_move() != 0){
                printf("ERROR! Func: AI_move\n");
                return 1;
            }
        }
        if(per%2 == 1){
            if(level == 1) printf("Режим: сложный | ");
            if(level == 0) printf("Режим: лёгкий | ");
            printf("Вы играете за: '%c'\n",m);
            while(sub){
                co[0] = 4;
                co[1] = 4;
                printf("Куда хотите поставить '%c' (Чтобы выйти введите 5 5):",m);
                for(i=0;i<2;i++) scanf("%zu",&co[i]);
                fflush(stdin);
                if(co[0] == 5 && co[1] == 5){
                    user_br = false;
                    break;    
                }
                if(co[0] > 3 || co[1] > 3 || co[0] < 1 || co[1] < 1){
                    printf("Error!--Введено неверное значение.Попробуйте снова.\n\n");
                    continue;
                }
                if(a[co[0]-1][co[1]-1] != '_'){
                    printf("Данное поле уже занято значком:'%c'\n\n",a[co[0]-1][co[1]-1]);
                    continue;
                }
                else{
                    a[co[0]-1][co[1]-1] = m;
                    sub = FALSE;
                }
            }
            sub = TRUE;
        }
        for(i=0;i < 3;i++){
            for(j=0;j < 3;j++) printf(" %c ",a[i][j]);
            printf("\n");
        }
        if(scan(&flag,&br,&win) != 0){
            printf("ERROR! Func: scan, with argument %d\n",scan(&flag,&br,&win));
            return 1;
        }
        per += 1; 
    }
    if(!user_br) printf("Игра завершена");
    if(!br && flag) printf("Победителя нет");
    if(!flag) printf("Победил '%c'",win);
    return 0;
}

int scan(bool *flag,bool *br,char *win){
    if(flag == NULL) return 1;
    if(br == NULL) return 2;
    if(win == NULL) return 3;

    *br = false;
    for(size_t i = 0; i < 3; i++){
        if(scan_line(a[i][0],a[i][1],a[i][2])){//для строк
            *win = a[i][0];
            *flag = false;
        }
    }
    for(size_t i = 0; i < 3; i++){
        if(scan_line(a[0][i],a[1][i],a[2][i])){//для столбцов
            *win = a[0][i];
            *flag = false;
        }
    }
    if(scan_line(a[0][0],a[1][1],a[2][2])){
        *win = a[0][0];
        *flag = false;
    }
    if(scan_line(a[0][2],a[1][1],a[2][0])){
        *win = a[2][0];
        *flag = false;
    }
    for(size_t i = 0; i < 3; i++){
        for(size_t j = 0; j < 3; j++){
            if(a[i][j] == '_'){
                *br = true;
            }
        }
    }
    return 0;
}

bool scan_line(char el1,char el2,char el3){
    bool result = false;
    if(el1 == el2 && el2 == el3 && el3 != '_') result = true;
    return result;
}

int AI_move(){
    int result;
    if(m == 'o'){
        value = 'x';
        if(level == 1){
            if(per == 0){
                a[0][0] = 'x';
                return 0;
            }
            if(per == 2){
                if(a[2][0] == '_') a[2][0] = value;
                else a[2][2] = value;
                return 0;
            }
        }
    }
    if(m == 'x'){
        value = 'o';
        if(level == 1){
            if(per == 2){
                if(a[1][1] == '_') a[1][1] = value;
                else a[0][0] = value;
                return 0;
            }
        }
    }
    if((per >= 4 && level == 1) || (level == 0)){
        //обрабатываем критическую ситуацию(2 value)
        for(size_t i = 0; i < 3; i++){//для строк
            result = scan_to_win(&a[i][0],&a[i][1],&a[i][2],value);
            if(result && result != 1 && result != 2 && result != 3){
                return 0;
            }
            else if(result == 1 || result == 2 || result == 3){
                printf("ERROR! Func: scan_to_win\n");
                return 1;
            }
        }
        
        for(size_t i = 0; i < 3; i++){//для столбцов
            result = scan_to_win(&a[0][i],&a[1][i],&a[2][i],value);
            if(result && result != 1 && result != 2 && result != 3){
                return 0;
            }
            else if(result == 1 || result == 2 || result == 3){
                printf("ERROR! Func: scan_to_win\n");
                return 1;
            }
        }

        result = scan_to_win(&a[0][0],&a[1][1],&a[2][2],value);
        if(result && result != 1 && result != 2 && result != 3){
            return 0;
        }
        else if(result == 1 || result == 2 || result == 3){
            printf("ERROR! Func: scan_to_win\n");
            return 1;
        }

        result = scan_to_win(&a[0][2],&a[1][1],&a[2][0],value);
        if(result && result != 1 && result != 2 && result != 3){
            return 0;
        }
        else if(result == 1 || result == 2 || result == 3){
            printf("ERROR! Func: scan_to_win\n");
            return 1;
        }
        //обрабатываем критическую ситуацию(2 m)
        for(size_t i = 0; i < 3; i++){//для строк
            result = scan_to_win(&a[i][0],&a[i][1],&a[i][2],m);
            if(result && result != 1 && result != 2 && result != 3){
                return 0;
            }
            else if(result == 1 || result == 2 || result == 3){
                printf("ERROR! Func: scan_to_win\n");
                return 1;
            }
        }
        
        for(size_t i = 0; i < 3; i++){//для столбцов
            result = scan_to_win(&a[0][i],&a[1][i],&a[2][i],m);
            if(result && result != 1 && result != 2 && result != 3){
                return 0;
            }
            else if(result == 1 || result == 2 || result == 3){
                printf("ERROR! Func: scan_to_win\n");
                return 1;
            }
        }

        result = scan_to_win(&a[0][0],&a[1][1],&a[2][2],m);
        if(result && result != 1 && result != 2 && result != 3){
            return 0;
        }
        else if(result == 1 || result == 2 || result == 3){
            printf("ERROR! Func: scan_to_win\n");
            return 1;
        }

        result = scan_to_win(&a[0][2],&a[1][1],&a[2][0],m);
        if(result && result != 1 && result != 2 && result != 3){
            return 0;
        }
        else if(result == 1 || result == 2 || result == 3){
            printf("ERROR! Func: scan_to_win\n");
            return 1;
        }

        //проверяем случай когда 1 value и остальное пусто
        for(size_t i = 0; i < 3; i++){//для строк
            result = scan_to_move(&a[i][0],&a[i][1],&a[i][2]);
            if(result && result != 1 && result != 2 && result != 3){
                return 0;
            }
            else if(result == 1 || result == 2 || result == 3){
                printf("ERROR! Func: scan_to_move\n");
                return 1;
            }
        }

        for(size_t i = 0; i < 3; i++){//для столбцов
            result = scan_to_move(&a[0][i],&a[1][i],&a[2][i]); 
            if(result && result != 1 && result != 2 && result != 3){
                return 0;
            }
            else if(result == 1 || result == 2 || result == 3){
                printf("ERROR! Func: scan_to_move\n");
                return 1;
            }
        }

        result = scan_to_move(&a[0][0],&a[1][1],&a[2][2]);
        if(result && result != 1 && result != 2 && result != 3){
            return 0;
        }
        else if(result == 1 || result == 2 || result == 3){
            printf("ERROR! Func: scan_to_move\n");
            return 1;
        }

        result = scan_to_move(&a[0][2],&a[1][1],&a[2][0]);
        if(result && result != 1 && result != 2 && result != 3){
            return 0;
        }
        else if(result == 1 || result == 2 || result == 3){
            printf("ERROR! Func: scan_to_move\n");
            return 1;
        }
        //во всех остальных случаях значение ставится рандомно
        srand(time(NULL));
        while(true){
            size_t i = rand()%3, j = rand()%3;
            if(a[i][j] == '_'){
                a[i][j] = value;
                return 0;
            }
        }
    }
    return 1;
}

int scan_to_win(char *el1,char *el2,char *el3,char val){
    if(el1 == NULL) return 1;
    if(el2 == NULL) return 2;
    if(el3 == NULL) return 3;
    int result = 0;
	if (*el1 == *el2 && *el1 == val && *el3 == '_'){
		*el3 = value;
		result = 5;
	}
	if (*el1 == *el3 && *el1 == val && *el2 == '_'){
		*el2 = value;
		result = 5;
    }
	if (*el2 == *el3 && *el2 == val && *el1 == '_'){
		*el1 = value;
		result = 5;
    }
    return result;
}

int scan_to_move(char *el1, char *el2, char *el3){
    if(el1 == NULL) return 1;
    if(el2 == NULL) return 2;
    if(el3 == NULL) return 3;
	int result = 0;
	size_t i = rand()%2;
	srand(time(NULL));
	if (*el1 == value && *el2 == '_' && *el3 == '_') {
		if (i == 0) *el2 = value;
		else *el3 = value;
		result = 4;
	}
	if (*el2 == value && *el1 == '_' && *el3 == '_') {
		if (i == 0) *el1 = value;
		else *el3 = value;
		result = 4;
	}
	if (*el3 == value && *el1 == '_' && *el2 == '_') {
		if (i == 0) *el1 = value;
		else *el2 = value;
		result = 4;
	}
	return result;
}